### ISSUE TITLE
Pixelorama crashes when cropping image multiple times

### ISSUE THREAD

**Issue by qarmin** [Orama-Interactive/Pixelorama#421]
**Pixelorama version:**
0.8.2 stable

**Issue description:**
When I multiple times crop image, then sometimes Pixelorama crash or start to shows errors

Looks that Godot tries to use memory after free (Log from Godot compiled with sanitizers support, but crash should show similar bactrace)
```
==227768==ERROR: AddressSanitizer: heap-use-after-free on address 0x629000118210 at pc 0x00000e700639 bp 0x7ffeeee90160 sp 0x7ffeeee90150
READ of size 1 at 0x629000118210 thread T0
    #0 0xe700638 in Image::get_pixel(int, int) const core/image.cpp:2450
    #1 0xe6ec49c in Image::get_used_rect() const core/image.cpp:2052
    #2 0x97f0b43 in MethodBind0RC<Rect2>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:593
    #3 0xe82ec03 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:919
    #4 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #5 0x19fa820 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1088
    #6 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #7 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #8 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #9 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #10 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #11 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #12 0xe838a64 in Object::emit_signal(StringName const&, Variant const**, int) core/object.cpp:1246
    #13 0xe83aa9a in Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) core/object.cpp:1303
    #14 0x9e5ce2f in PopupMenu::activate_item(int) scene/gui/popup_menu.cpp:1198
    #15 0x9e2ea17 in PopupMenu::_gui_input(Ref<InputEvent> const&) scene/gui/popup_menu.cpp:365
    #16 0x235587c in MethodBind1<Ref<InputEvent> const&>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:775
    #17 0xe82a82c in Object::call_multilevel(StringName const&, Variant const**, int) core/object.cpp:761
    #18 0xe82d928 in Object::call_multilevel(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) core/object.cpp:861
    #19 0x96fef36 in Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) scene/main/viewport.cpp:1688
    #20 0x970e5d2 in Viewport::_gui_input_event(Ref<InputEvent>) scene/main/viewport.cpp:2068
    #21 0x97333c4 in Viewport::input(Ref<InputEvent> const&) scene/main/viewport.cpp:2858
    #22 0x96f2b23 in Viewport::_vp_input(Ref<InputEvent> const&) scene/main/viewport.cpp:1453
    #23 0x235587c in MethodBind1<Ref<InputEvent> const&>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:775
    #24 0xe82ec03 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:919
    #25 0xe82d359 in Object::call(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) core/object.cpp:845
    #26 0x95df705 in SceneTree::call_group_flags(unsigned int, StringName const&, StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) scene/main/scene_tree.cpp:275
    #27 0x95e4baf in SceneTree::input_event(Ref<InputEvent> const&) scene/main/scene_tree.cpp:431
    #28 0x14a29eb in InputDefault::_parse_input_event_impl(Ref<InputEvent> const&, bool) main/input_default.cpp:442
    #29 0x1498141 in InputDefault::parse_input_event(Ref<InputEvent> const&) main/input_default.cpp:259
    #30 0x14ab0c3 in InputDefault::flush_accumulated_events() main/input_default.cpp:678
    #31 0x1421188 in OS_X11::process_xevents() platform/x11/os_x11.cpp:2865
    #32 0x1437d2f in OS_X11::run() platform/x11/os_x11.cpp:3604
    #33 0x13a5b56 in main platform/x11/godot_x11.cpp:56
    #34 0x7ff2f08be0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #35 0x13a576d in _start (/usr/bin/godots+0x13a576d)

0x629000118210 is located 16 bytes inside of 16400-byte region [0x629000118200,0x62900011c210)
freed by thread T0 here:
    #0 0x7ff2f1d831b7 in __interceptor_free (/lib/x86_64-linux-gnu/libasan.so.6+0xb01b7)
    #1 0xed69fa4 in Memory::free_static(void*, bool) core/os/memory.cpp:178
    #2 0x1454ca2 in PoolVector<unsigned char>::_unreference() core/pool_vector.h:241
    #3 0x2da618e in PoolVector<unsigned char>::_reference(PoolVector<unsigned char> const&) core/pool_vector.h:191
    #4 0x2da5f61 in PoolVector<unsigned char>::operator=(PoolVector<unsigned char> const&) core/pool_vector.h:463
    #5 0xe6e20fa in Image::create(int, int, bool, Image::Format, PoolVector<unsigned char> const&) core/image.cpp:1619
    #6 0xe6fcbda in Image::_set_data(Dictionary const&) core/image.cpp:2380
    #7 0x2011898 in MethodBind1<Dictionary const&>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:775
    #8 0xe5c3bee in ClassDB::set_property(Object*, StringName const&, Variant const&, bool*) core/class_db.cpp:1049
    #9 0xe81f664 in Object::set(StringName const&, Variant const&, bool*) core/object.cpp:423
    #10 0xe9d7d13 in UndoRedo::_process_operation_list(List<UndoRedo::Operation, DefaultAllocator>::Element*) core/undo_redo.cpp:311
    #11 0xe9d91d7 in UndoRedo::redo() core/undo_redo.cpp:337
    #12 0xe9d5bd6 in UndoRedo::commit_action() core/undo_redo.cpp:259
    #13 0x1760003 in MethodBind0::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:59
    #14 0xe82ec03 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:919
    #15 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #16 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #17 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #18 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #19 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #20 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #21 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #22 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #23 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #24 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #25 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #26 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #27 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #28 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #29 0x182a30e in GDScriptInstance::call_multilevel(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1254

previously allocated by thread T0 here:
    #0 0x7ff2f1d83517 in malloc (/lib/x86_64-linux-gnu/libasan.so.6+0xb0517)
    #1 0xed69097 in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:82
    #2 0x16258b1 in PoolVector<unsigned char>::_copy_on_write() core/pool_vector.h:130
    #3 0x161f585 in PoolVector<unsigned char>::write() core/pool_vector.h:355
    #4 0xe6fe400 in Image::lock() core/image.cpp:2397
    #5 0x1760003 in MethodBind0::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:59
    #6 0xe82ec03 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:919
    #7 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #8 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #9 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #10 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #11 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #12 0x19fa820 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1088
    #13 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #14 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #15 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #16 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #17 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #18 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #19 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #20 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #21 0x1829dfb in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1238
    #22 0xe82e773 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:898
    #23 0xeaadfcf in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1129
    #24 0x19fa8a1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1091
    #25 0x182a30e in GDScriptInstance::call_multilevel(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1254
    #26 0xe82a501 in Object::call_multilevel(StringName const&, Variant const**, int) core/object.cpp:753
    #27 0x95f8b6b in SceneTree::_call_input_pause(StringName const&, StringName const&, Ref<InputEvent> const&) scene/main/scene_tree.cpp:945
    #28 0x973312d in Viewport::input(Ref<InputEvent> const&) scene/main/viewport.cpp:2854
    #29 0xa5332b0 in ViewportContainer::_input(Ref<InputEvent> const&) scene/gui/viewport_container.cpp:166

SUMMARY: AddressSanitizer: heap-use-after-free core/image.cpp:2450 in Image::get_pixel(int, int) const
```

**Steps to reproduce:**
1. Created big image (> 512px512px)
2. Draw anywhere small circle or something else
3. At the to click at Image -> Crop Image
4. Repeat from 1 point if still doesn't crash

**Comment by MidnightGreenlight** [comment]
When reproducing the bug, an error saying "resize: Can't resize PoolVector if locked" appears in the debug console, which ends up causing a crash.
This seems to be related to issue #331.

**Comment by qarmin** [comment]
Image class is the most bugged class I have ever seen in Godot.
I created in Godot several bug reports like e.g https://github.com/godotengine/godot/issues/46485 which happens mostly when image is locked.
After fixing this issues I think, that it will fix Pixelorama crashes

**Comment by qarmin** [comment]
Looks that this is fixed with Godot 3.4

**Comment by qarmin** [comment]
I tested wrong commit.
I need to wait for cherrypicking https://github.com/godotengine/godot/pull/50782

**Comment by OverloadedOrama** [comment]
This is most likely fixed now in Pixelorama by 0d5e9ce48aa9f56dd22236d3ea6a8f54a24d8e92, so I will close the issue. If it's not actually fixed, feel free to re-open it.

Godot's image class behaves oddly when images are being kept locked. Not sure if this is something that can get fixed from Godot's side or we should be careful and unlock images after locking them.
